experimental/dds/tree2/src/feature-libraries/memoizedIdRangeAllocator.ts

```diff
@@ -42,55 +59,75 @@ export interface IdRange {
  */
 export const MemoizedIdRangeAllocator = {
 	fromNextId(nextId: number = 0): MemoizedIdRangeAllocator {
-		let _nextId = nextId;
 		const rangeMap: Map<RevisionTag | undefined, RangeMap<number>> = new Map();
-		return (key: string | number | undefined, startId: number, length?: number): IdRange[] => {
-			let count = length ?? 1;
-			const out: IdRange[] = [];
-			const ranges = getOrAddEmptyToMap(rangeMap, key);
-			let currId = startId;
-			while (count > 0) {
-				const firstRange = getFirstFromRangeMap(ranges, currId, count);
-				if (firstRange === undefined) {
-					const newId = _nextId;
-					_nextId += count;
+		return {
+			nextId,
+			allocate(
+				key: string | number | undefined,
+				startId: number,
+				length?: number,
+			): IdRange[] {
+				let count = length ?? 1;
+				const out: IdRange[] = [];
+				const ranges = getOrAddEmptyToMap(rangeMap, key);
+				let currId = startId;
+				while (count > 0) {
+					const firstRange = getFirstFromRangeMap(ranges, currId, count);
+					if (firstRange === undefined) {
+						const newId = this.nextId;
+						this.nextId += count;
+						setInRangeMap(ranges, currId, count, newId);
+						out.push({ first: brand(newId), count });
+						count = 0;
+					} else {
+						const idRange: Mutable<IdRange> = {
+							first: brand(firstRange.value),
+							count: firstRange.length,
+						};
+						if (currId < firstRange.start) {
+							const countToAdd = firstRange.start - currId;
+							setInRangeMap(ranges, currId, countToAdd, this.nextId);
+							out.push({ first: brand(this.nextId), count: countToAdd });
+							this.nextId += countToAdd;
+							currId += countToAdd;
+							count -= countToAdd;
+						} else if (firstRange.start < currId) {
+							const countToTrim = currId - firstRange.start;
+							idRange.first = brand((idRange.first as number) + countToTrim);
+							idRange.count -= countToTrim;
+						}
+						if (idRange.count > count) {
+							idRange.count = count;
+						} else if (
+							idRange.count < count &&
+							firstRange.value + firstRange.length === this.nextId
+						) {
+							// The existing range can be extended
+							this.nextId += count - idRange.count;
+							firstRange.length = count;
+							idRange.count = count;
+						}
+						out.push(idRange);
+						count -= idRange.count;
+						currId += idRange.count;
+					}
+				}
+				return out;
+			},
+			mint(length?: number): IdRange[] {
```

Comment:
I think this can just be:
```
mint(length?: number): ChangesetLocalId {
	count count = length ?? 1;
	const out = this.nextId;
	this.nextId += count;
	return out;
}
```
This is because there's no way for the caller to ask for the same IDs again.